js/flexbuffers.js

```diff
@@ -0,0 +1,1015 @@
+      view.setUint16(offset, value, true);
+    } else if (width === flexbuffers.BitWidth.WIDTH32) {
+      view.setUint32(offset, value, true);
+    } else if (width === flexbuffers.BitWidth.WIDTH64) {
+      view.setBigUint64(offset, BigInt(value), true);
+    } else {
+      throw `Unexpected width: ${width} for value: ${value}`;
+    }
+  }
+
+  function writeInt(value, byteWidth) {
+    const newOffset = computeOffset(byteWidth);
+    pushInt(value, flexbuffers.BitWidth.fromByteWidth(byteWidth));
+    offset = newOffset;
+  }
+
+  function writeBlob(arrayBuffer) {
+    const length = arrayBuffer.byteLength;
+    const bitWidth = flexbuffers.BitWidth.width(length);
+    const byteWidth = align(bitWidth);
+    writeInt(length, byteWidth);
+    const blobOffset = offset;
+    const newOffset = computeOffset(length);
+    new Uint8Array(buffer).set(new Uint8Array(arrayBuffer), blobOffset);
+    stack.push(offsetStackValue(blobOffset, flexbuffers.ValueType.BLOB, bitWidth));
+    offset = newOffset;
+  }
+
+  function writeString(str) {
+    if (stringCache.hasOwnProperty(str)) {
+      stack.push(stringCache[str]);
+      return;
+    }
+    const utf8 = toUTF8Array(str);
+    const length = utf8.length;
+    const bitWidth = flexbuffers.BitWidth.width(length);
+    const byteWidth = align(bitWidth);
+    writeInt(length, byteWidth);
+    const stringOffset = offset;
+    const newOffset = computeOffset(length + 1);
+    new Uint8Array(buffer).set(utf8, stringOffset);
+    const stackValue = offsetStackValue(stringOffset, flexbuffers.ValueType.STRING, bitWidth);
+    stack.push(stackValue);
+    stringCache[str] = stackValue;
+    offset = newOffset;
+  }
+
+  function writeKey(str) {
+    if (keyCache.hasOwnProperty(str)) {
+      stack.push(keyCache[str]);
+      return;
+    }
+    const utf8 = toUTF8Array(str);
+    const length = utf8.length;
+    const newOffset = computeOffset(length + 1);
+    new Uint8Array(buffer).set(utf8, offset);
+    const stackValue = offsetStackValue(offset, flexbuffers.ValueType.KEY, flexbuffers.BitWidth.WIDTH8);
+    stack.push(stackValue);
+    keyCache[str] = stackValue;
+    offset = newOffset;
+  }
+
+  function writeStackValue(value, byteWidth) {
+    const newOffset = computeOffset(byteWidth);
+    if (value.isOffset) {
+      const relativeOffset = offset - value.offset;
+      if (byteWidth === 8 || BigInt(relativeOffset) < (1n << BigInt(byteWidth * 8))) {
+        writeInt(relativeOffset, byteWidth);
+      } else {
+        throw `Unexpected size ${byteWidth}. This might be a bug. Please create an issue https://github.com/google/flatbuffers/issues/new`
+      }
+    } else {
+      value.writeToBuffer(byteWidth);
+    }
+    offset = newOffset;
+  }
+
+  function integrityCheckOnValueAddition() {
+    if (finished) {
+      throw "Adding values after finish is prohibited";
+    }
+    if (stackPointers.length !== 0 && stackPointers[stackPointers.length - 1].isVector === false) {
```

Comment:
why not do this check as a map is being serialized, which is much cheaper. See C++ implementation.

Reply:
Because IMHO it is better to keep cause and effect close together. This way the exception is thrown when I add the value so I will know how to fix it instantly. If the exception is thrown when the map is being serialized, I know I did something wrong, but I have no idea what exactly so I will have to manually go through my builder code and search for one or multiple bugs.

Reply:
for errors that are not likely to be frequent, I'd rather have a focus on performance.